File: desktop/packet_processor.py
import struct
import json
from enum import IntEnum


class PacketType(IntEnum):
    JSON = 0,
    JPG = 1,
    ACK = 2


class PacketProcessor:
    MAX_PACKET_SIZE = 1400
    MAX_PAYLOAD_SIZE = MAX_PACKET_SIZE - 3

    def __init__(self):
        self.send_packet_number = 0
        self.packets = []
        self.is_crypted = False
        self.send_partial_offset = 0
        self.receive_partial_chunk = b''
        self.last_received_packet_number = 0
        self.last_received_time = 0

    def pack(self) -> bytes:
        if len(self.packets) == 0:
            return

        self.send_packet_number = self.get_next_packet_number()

        fit_it = 0
        sz = 0
        for p in self.packets:
            chunk_len = len(p[1]) + 5
            if fit_it == 0:
                chunk_len -= self.send_partial_offset

            if fit_it != 0 and sz + chunk_len > self.MAX_PAYLOAD_SIZE:
                break
            fit_it += 1
            sz += chunk_len

        first_chunk_over_size = sz > self.MAX_PAYLOAD_SIZE
        if first_chunk_over_size:
            sz = self.MAX_PAYLOAD_SIZE

        next_partial_offset = self.send_partial_offset

        payload = bytearray(sz)
        off = 0
        for i in range(0, fit_it):
            p = self.packets[i]
            chunk_len = len(p[1])

            payload[off:off+4] = struct.pack("!I", chunk_len)
            payload[off+4] = p[0]

            chunk_offset = 0

            if i == 0:
                chunk_offset = self.send_partial_offset
                chunk_len -= self.send_partial_offset

            data_off = off + 5
            if data_off + chunk_len > sz:
                chunk_len = sz - data_off
                next_partial_offset += chunk_len
            else:
                next_partial_offset = 0

            payload[data_off:data_off+chunk_len] = p[1][chunk_offset:chunk_offset+chunk_len]
            off += chunk_len + 5

        if self.is_crypted:
            payload = self.crypt(payload)

        sz = len(payload)
        ret = bytearray(sz + 3)
        ret[:2] = struct.pack("!H", self.send_packet_number)
        ret[2] = 1 if self.send_partial_offset != 0 else 0
        ret[3:] = payload

        self.send_partial_offset = next_partial_offset

        if not first_chunk_over_size:
            self.packets = self.packets[fit_it:]

        return ret

    def get_next_packet_number(self):
        ret = self.send_packet_number + 1
        ret %= 65536
        return ret

    def get_next_packet_number(self):
        ret = self.send_packet_number + 1
        ret %= 65536
        return ret

    @staticmethod
    def pack_chunk(data, tp: PacketType):
        return tp, data

    def pack_json(self, js) -> bytes:
        return self.pack_chunk(json.dumps(js).encode('utf-8'), PacketType.JSON)

    def pack_ack(self, received_packet_number) -> bytes:
        bts = struct.pack("!H", received_packet_number)
        return self.pack_chunk(bts, PacketType.ACK)

    @staticmethod
    def get_packet_number(packet: bytes):
        total_len = len(packet)
        if total_len < 2:
            return None

        packet_number, = struct.unpack("!H", packet[0:2])
        return packet_number

    def is_packet_too_old(self, pack_n):
        return self.last_received_packet_number >= pack_n and \
                not (pack_n < 64 and self.last_received_packet_number > 65536 - 64)

    def parse(self, packet: bytes):
        if len(packet) < 3:
            return False

        packet_number, = struct.unpack("!H", packet[0:2])
        is_next_partial = packet[2] != 0
        i = 3

        if is_next_partial and packet_number != self.last_received_packet_number + 1:
            return False

        if self.is_crypted:
            payload = self.decrypt(packet[i:], packet_number)
            if payload is None:
                return False
            return self.iparse(payload, 0, is_next_partial)
        else:
            return self.iparse(packet, i, is_next_partial)

    def iparse(self, packet, offset, is_next_partial):
        if not is_next_partial:
            self.receive_partial_chunk = b''
        received_partial_len = len(self.receive_partial_chunk)

        total_len = len(packet)
        i = offset
        item = 0
        while i < total_len:
            if i + 5 > total_len:
                return False
            chunk_len, = struct.unpack("!I", packet[i:i+4])

            if item == 0:
                if chunk_len <= received_partial_len:
                    return False
                chunk_len -= received_partial_len

            next_i = i+5+chunk_len
            if next_i > total_len and item != 0:
                return False

            i = next_i
            item += 1

        i = offset
        item = 0
        #print(f'iparse()2: {is_next_partial=} {received_partial_len=} {total_len=} {i=}')
        while i+5 <= total_len:
            chunk_len, = struct.unpack("!I", packet[i:i+4])

            if item == 0:
                chunk_len -= received_partial_len

            #print(f'iparse()3: {chunk_len=}')

            next_i = i+5+chunk_len
            if next_i > total_len and item != 0:
                break

            chunk = packet[i + 5:i + 5 + chunk_len]
            chunk_type = packet[i+4]

            if item == 0:
                if next_i > total_len:
                    self.receive_partial_chunk += chunk
                    #print(f'iparse()5: {len(self.receive_partial_chunk)=}')
                    return True
                chunk = self.receive_partial_chunk + chunk
                self.receive_partial_chunk = b''
                received_partial_len = 0

            try:
                if chunk_type == PacketType.JSON:
                    js = json.loads(chunk)
                    self.process_json(js)
                elif chunk_type == PacketType.JPG:
                    self.process_jpeg(chunk)
                elif chunk_type == PacketType.ACK:
                    ack_packet_number, = struct.unpack("!H", chunk)
                    self.process_ack(ack_packet_number)

            except Exception as e:
                print(e)

            i = next_i

        return True

    def crypt(self, bts):
        return bts

    def decrypt(self, bts, packet_number):
        return bts

    def process_json(self, js):
        pass

    def process_jpeg(self, bts):
        pass

    def process_ack(self, ack_packet_number):
        pass

File: android/data/src/main/java/com/robocart/data/network/PacketProcessor.kt
package com.robocart.data.network

import org.json.JSONArray
import org.json.JSONObject
import org.json.JSONTokener
import java.nio.charset.StandardCharsets
import kotlin.math.min

open class PacketProcessor {
    companion object {
        const val MAX_PACKET_SIZE = 1400
        const val MAX_PAYLOAD_SIZE = MAX_PACKET_SIZE - 3
    }

    data class PacketChunk(val type: PacketType, val payload: ByteArray)

    protected val packets = ArrayDeque<PacketChunk>()
    private var sendPacketNumber = 0
    protected var isCrypted = false
    private var sendPartialOffset = 0
    private var receivePartialChunk = ByteArray(0)
    protected var lastReceivedPacketNumber = 0
    protected var lastReceivedTimeSec = 0L

    protected fun pack(): ByteArray? {
        if (packets.isEmpty()) {
            return null
        }

        sendPacketNumber = getNextPacketNumber()
        var fitIt = 0
        var totalPayloadSize = 0
        for (chunk in packets) {
            var chunkLen = chunk.payload.size + 5
            if (fitIt == 0) {
                chunkLen -= sendPartialOffset
            }
            if (fitIt != 0 && totalPayloadSize + chunkLen > MAX_PAYLOAD_SIZE) {
                break
            }
            fitIt++
            totalPayloadSize += chunkLen
        }

        val firstChunkOversized = totalPayloadSize > MAX_PAYLOAD_SIZE
        if (firstChunkOversized) {
            totalPayloadSize = MAX_PAYLOAD_SIZE
        }

        var nextPartialOffset = sendPartialOffset
        val payload = ByteArray(totalPayloadSize)
        var offset = 0
        for (i in 0 until fitIt) {
            val packet = packets.elementAt(i)
            var chunkLen = packet.payload.size
            writeUInt32(payload, offset, chunkLen.toUInt())
            payload[offset + 4] = packet.type.wireValue.toByte()

            var chunkOffset = 0
            if (i == 0) {
                chunkOffset = sendPartialOffset
                chunkLen -= sendPartialOffset
            }

            val dataOffset = offset + 5
            if (dataOffset + chunkLen > totalPayloadSize) {
                chunkLen = totalPayloadSize - dataOffset
                nextPartialOffset += chunkLen
            } else {
                nextPartialOffset = 0
            }

            System.arraycopy(packet.payload, chunkOffset, payload, dataOffset, chunkLen)
            offset += chunkLen + 5
        }

        val finalPayload = if (isCrypted) crypt(payload) else payload
        val packet = ByteArray(finalPayload.size + 3)
        packet[0] = ((sendPacketNumber ushr 8) and 0xFF).toByte()
        packet[1] = (sendPacketNumber and 0xFF).toByte()
        packet[2] = if (sendPartialOffset != 0) 1 else 0
        System.arraycopy(finalPayload, 0, packet, 3, finalPayload.size)

        sendPartialOffset = nextPartialOffset
        if (!firstChunkOversized) {
            repeat(fitIt) { packets.removeFirst() }
        }
        return packet
    }

    fun packChunk(data: ByteArray, type: PacketType): PacketChunk = PacketChunk(type, data)

    fun packJson(json: String): PacketChunk = packChunk(json.toByteArray(StandardCharsets.UTF_8), PacketType.JSON)

    fun packJson(payload: Map<String, Any?>): PacketChunk {
        val jsonObject = JSONObject()
        payload.forEach { (key, value) ->
            jsonObject.put(key, toJsonValue(value))
        }
        return packJson(jsonObject.toString())
    }

    fun packAck(receivedPacketNumber: Int): PacketChunk {
        val bytes = byteArrayOf(
            ((receivedPacketNumber ushr 8) and 0xFF).toByte(),
            (receivedPacketNumber and 0xFF).toByte(),
        )
        return packChunk(bytes, PacketType.ACK)
    }

    fun enqueueChunk(chunk: PacketChunk) {
        packets.addLast(chunk)
    }

    fun enqueueJson(payload: Map<String, Any?>) {
        enqueueChunk(packJson(payload))
    }

    fun getPacketNumber(packet: ByteArray): Int? {
        if (packet.size < 2) {
            return null
        }
        return ((packet[0].toInt() and 0xFF) shl 8) or (packet[1].toInt() and 0xFF)
    }

    fun isPacketTooOld(packetNumber: Int): Boolean {
        return lastReceivedPacketNumber >= packetNumber &&
            !(packetNumber < 64 && lastReceivedPacketNumber > 65536 - 64)
    }

    fun parse(packet: ByteArray): Boolean {
        if (packet.size < 3) {
            return false
        }

        val packetNumber = getPacketNumber(packet) ?: return false
        val isNextPartial = packet[2].toInt() != 0
        if (isNextPartial && packetNumber != ((lastReceivedPacketNumber + 1) and 0xFFFF)) {
            return false
        }

        return if (isCrypted) {
            val payload = decrypt(packet.copyOfRange(3, packet.size), packetNumber) ?: return false
            parseInternal(payload, 0, isNextPartial)
        } else {
            parseInternal(packet, 3, isNextPartial)
        }
    }

    private fun parseInternal(packet: ByteArray, offset: Int, isNextPartial: Boolean): Boolean {
        if (!isNextPartial) {
            receivePartialChunk = ByteArray(0)
        }
        var receivedPartialLen = receivePartialChunk.size
        val totalLen = packet.size

        var i = offset
        var item = 0
        while (i < totalLen) {
            if (i + 5 > totalLen) {
                return false
            }
            var chunkLen = readUInt32(packet, i).toInt()
            if (item == 0) {
                if (chunkLen <= receivedPartialLen) {
                    return false
                }
                chunkLen -= receivedPartialLen
            }
            val nextI = i + 5 + chunkLen
            if (nextI > totalLen && item != 0) {
                // The trailing chunk may legitimately be partial; it will be completed by the next packet.
                break
            }
            i = nextI
            item++
        }

        i = offset
        item = 0
        while (i + 5 <= totalLen) {
            var chunkLen = readUInt32(packet, i).toInt()
            if (item == 0) {
                chunkLen -= receivedPartialLen
            }

            val nextI = i + 5 + chunkLen
            if (nextI > totalLen && item != 0) {
                break
            }

            var chunk = packet.copyOfRange(i + 5, min(i + 5 + chunkLen, totalLen))
            val chunkType = packet[i + 4].toInt() and 0xFF

            if (nextI > totalLen) {
                receivePartialChunk += chunk
                return true
            }

            if (item == 0) {
                chunk = receivePartialChunk + chunk
                receivePartialChunk = ByteArray(0)
                receivedPartialLen = 0
            }

            try {
                when (PacketType.fromWireValue(chunkType)) {
                    PacketType.JSON -> processJson(JSONTokener(String(chunk, StandardCharsets.UTF_8)).nextValue())
                    PacketType.JPG -> processJpeg(chunk)
                    PacketType.ACK -> {
                        if (chunk.size >= 2) {
                            val ackNumber = ((chunk[0].toInt() and 0xFF) shl 8) or (chunk[1].toInt() and 0xFF)
                            processAck(ackNumber)
                        }
                    }
                    null -> Unit
                }
            } catch (_: Exception) {
                // Keep parser resilient to malformed chunks, matching Python behavior.
            }
            i = nextI
            item++
        }
        return true
    }

    private fun getNextPacketNumber(): Int = (sendPacketNumber + 1) % 65536

    private fun writeUInt32(buffer: ByteArray, offset: Int, value: UInt) {
        buffer[offset] = ((value.toLong() ushr 24) and 0xFF).toByte()
        buffer[offset + 1] = ((value.toLong() ushr 16) and 0xFF).toByte()
        buffer[offset + 2] = ((value.toLong() ushr 8) and 0xFF).toByte()
        buffer[offset + 3] = (value.toLong() and 0xFF).toByte()
    }

    private fun readUInt32(buffer: ByteArray, offset: Int): UInt {
        val a = buffer[offset].toUByte().toUInt() shl 24
        val b = buffer[offset + 1].toUByte().toUInt() shl 16
        val c = buffer[offset + 2].toUByte().toUInt() shl 8
        val d = buffer[offset + 3].toUByte().toUInt()
        return a or b or c or d
    }

    protected open fun crypt(bytes: ByteArray): ByteArray = bytes

    protected open fun decrypt(bytes: ByteArray, packetNumber: Int): ByteArray? = bytes

    protected open fun processJson(payload: Any?) = Unit

    protected open fun processJpeg(payload: ByteArray) = Unit

    protected open fun processAck(ackPacketNumber: Int) = Unit

    private fun toJsonValue(value: Any?): Any? {
        return when (value) {
            null -> JSONObject.NULL
            is Map<*, *> -> {
                val obj = JSONObject()
                value.forEach { (k, v) ->
                    if (k is String) {
                        obj.put(k, toJsonValue(v))
                    }
                }
                obj
            }
            is Iterable<*> -> {
                val arr = JSONArray()
                value.forEach { arr.put(toJsonValue(it)) }
                arr
            }
            else -> value
        }
    }
}
